test/metabase/api/card_test.clj

```diff
@@ -3658,8 +3723,27 @@
                 (is (not (mi/can-read? parent-card)))
                 (is (mi/can-read? allowed-collection))
                 (is (mi/can-read? child-card)))
-              (is (= [[1] [2]]
-                     (mt/rows (process-query-for-card child-card)))))))))))
+              (is (= :blocked
+                     (get-in (data-perms/permissions-for-user (mt/user->id :rasta)) [(mt/id) :perms/view-data]))
+                  "rasta is blocked for this table.")
+              (testing "Data perms prohibit running queries"
+                (is (thrown-with-msg?
+                     clojure.lang.ExceptionInfo
+                     #"You do not have permissions to run this query."
+                     (mt/rows (process-query-for-card child-card)))
+                    "Even if the user has can-write? on a Card, they should not be able to run it because they are blocked on Card's db")))
+            (testing "create-queries = query-builder-and-native is sufficent to run the query"
+              (mt/with-restored-data-perms!
+                (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/view-data :blocked)
+                (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/create-queries :query-builder-and-native)
```

Comment:
I was a little confused by this - if you look at the code for `set-database-permission!`, it will automatically set `:perms/view-data` to `:unrestricted` if you set `:perms/create-queries` to anything but `:no`. So i think what's actually going on here is that you need `:unrestricted` data access, and your `:perms/create-queries` permission doesn't matter.

Reply:
Fixed